firedrake/linear_solver.py

```diff
@@ -80,8 +82,14 @@ def __init__(self, A, P=None, solver_parameters=None,
             if P is not None:
                 transpose_nullspace._apply(self.P, transpose=True)
 
+        if near_nullspace is not None:
+            near_nullspace._apply(self.A)
+            if P is not None:
+                near_nullspace._apply(self.P)
```

Comment:
I think you need to pass `near=True` here.  Otherwise this will be treated as a nullspace, no?